src/App.tsx

```diff
@@ -1,26 +1,179 @@
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+
+  const filteredTodos = todos.filter(todo => {
+    switch (filter) {
+      case FilterType.All:
+        return true;
+      case FilterType.Completed:
+        return todo.completed;
+      case FilterType.Active:
+        return !todo.completed;
+      default:
+        return false;
+    }
+  });
+
+  const notCompletedTodosCount = todos.filter(todo => !todo.completed).length;
+  const allTodosAreCompleted = useMemo(
+    () => todos.every(todo => todo.completed),
+    [todos],
+  );
+
+  const inputNameRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage(ErrorType.LoadTodo);
+      });
+  }, []);
+
+  const handleFilterChange = (newFilter: FilterType) => {
+    setFilter(newFilter);
+  };
+
+  const handleDeleteTodo = async (todoId: number) => {
+    setLoadingTodoIds(prev => [...prev, todoId]);
+    try {
+      await deleteTodo(todoId);
+      setTodos(prev => prev.filter(todo => todo.id !== todoId));
+    } catch (error) {
+      setErrorMessage(ErrorType.DeleteTodo);
+      inputNameRef.current?.focus();
+    } finally {
+      setLoadingTodoIds(prev => prev.filter(id => id !== todoId));
+    }
+  };
+
+  const handleAddTodo = async (todoTitle: string) => {
+    setTempTodo({
+      id: 0,
+      title: todoTitle,
+      completed: false,
+      userId: USER_ID,
+    });
+    try {
+      const newTodo = await createTodo({ title: todoTitle, completed: false });
+
+      setTodos(prev => [...prev, newTodo]);
+    } catch (error) {
+      setErrorMessage(ErrorType.AddTodo);
+      inputNameRef.current?.focus();
+      throw error;
+    } finally {
+      setTempTodo(null);
+    }
+  };
+
+  const handleUpdateTodo = async (todoToUpdate: Todo) => {
+    setLoadingTodoIds(prev => [...prev, todoToUpdate.id]);
+    try {
+      const updatedTodo = await updateTodo(todoToUpdate);
+
+      setTodos(prev =>
+        prev.map(todo => (todo.id === updatedTodo.id ? updatedTodo : todo)),
+      );
+    } catch (err) {
+      setErrorMessage(ErrorType.UpdateTodo);
+      inputNameRef.current?.focus();
+      throw err;
+    } finally {
+      setLoadingTodoIds(prev => prev.filter(id => id !== todoToUpdate.id));
+    }
+  };
+
+  const handletoggleAll = async () => {
+    if (notCompletedTodosCount > 0) {
+      const activeTodos = todos.filter(todo => !todo.completed);
+
+      activeTodos.forEach(todo =>
+        handleUpdateTodo({ ...todo, completed: true }),
+      );
+    } else {
+      todos.forEach(todo => handleUpdateTodo({ ...todo, completed: false }));
+    }
+  };
+
+  const handleClearCompleted = async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    completedTodos.forEach(todo => {
+      handleDeleteTodo(todo.id);
+    });
```

Comment:
[Promise.all](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) will better in this case. U can try it